python/dgl/graphbolt/itemset.py

```diff
@@ -365,18 +367,38 @@ def __getitem__(self, index: Union[int, slice, Iterable[int]]):
         elif isinstance(index, Iterable):
             if not isinstance(index, torch.Tensor):
                 index = torch.tensor(index)
-            assert torch.all((index >= 0) & (index < self._length))
-            key_indices = (
-                torch.searchsorted(self._offsets, index, right=True) - 1
-            )
             data = {}
-            for key_id, key in enumerate(self._keys):
-                mask = (key_indices == key_id).nonzero().squeeze(1)
-                if len(mask) == 0:
-                    continue
-                data[key] = self._itemsets[key][
-                    index[mask] - self._offsets[key_id]
-                ]
+            if len(index) < self._threshold:
```

Comment:
I think we need benchmark before settings such a threshold only after looking at runtime complexity.

Reply:
You're right. I'll do it right away.

Reply:
Let's use K values 1 2 4 8 16 32 etc.

Reply:
This is a little bit complex, I think the optimal complexity should be O(N*logK), with additional 2 helper array: offsets = [0, 10, 30, 60], etypes = ["A", "B", "C", "D"]

Reply:
NVM, I see the points, to do the way I suggested, we need to implement our own C++ kernel with parallel optimization.

Reply:
Is any of the ops used in the new implementation single threaded?

Reply:
Seems no. Maybe sorting op itself is a bit slow even it's multi-threaded?

Reply:
How did tou verify?

Reply:
I have no idea. Did you find out anything from the benchmarking code?